types/src/transaction/authenticator.rs

```diff
@@ -737,16 +802,387 @@ impl fmt::Display for AuthenticationKey {
     }
 }
 
+#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+pub struct MultiKeyAuthenticator {
+    public_keys: MultiKey,
+    signatures: Vec<AnySignature>,
+    signatures_bitmap: aptos_bitvec::BitVec,
+}
+
+impl MultiKeyAuthenticator {
+    pub fn new(public_keys: MultiKey, signatures: Vec<(u8, AnySignature)>) -> Result<Self> {
+        ensure!(
+            public_keys.len() < (u8::MAX as usize),
+            "Too many public keys in MultiKeyAuthenticator."
+        );
+
+        let mut signatures_bitmap = aptos_bitvec::BitVec::with_num_bits(public_keys.len() as u16);
+        let mut any_signatures = vec![];
+
+        for (idx, signature) in signatures {
+            ensure!(
+                (idx as usize) < public_keys.len(),
+                "Signature index is out of public key range."
+            );
+            ensure!(
+                !signatures_bitmap.is_set(idx as u16),
+                "Duplicate signature index."
+            );
+            signatures_bitmap.set(idx as u16);
+            any_signatures.push(signature);
+        }
+
+        Ok(MultiKeyAuthenticator {
+            public_keys,
+            signatures: any_signatures,
+            signatures_bitmap,
+        })
+    }
+
+    pub fn public_keys(&self) -> &MultiKey {
+        &self.public_keys
+    }
+
+    pub fn signatures(&self) -> Vec<(u8, AnySignature)> {
+        let mut values = vec![];
+        for (idx, signature) in
+            std::iter::zip(self.signatures_bitmap.iter_ones(), self.signatures.iter())
+        {
+            values.push((idx as u8, signature.clone()));
+        }
+        values
+    }
+
+    pub fn verify<T: Serialize + CryptoHash>(&self, message: &T) -> Result<()> {
+        ensure!(
+            self.signatures_bitmap.count_ones() as usize == self.signatures.len(),
+            "Mismatch in signatures and signatures_bitmap"
+        );
+        println!(
+            "{} {}",
+            self.signatures.len(),
+            self.public_keys.signatures_required()
+        );
+        ensure!(
+            self.signatures.len() >= self.public_keys.signatures_required() as usize,
+            "Not enough signatures for verification"
+        );
+        for (idx, signature) in
+            std::iter::zip(self.signatures_bitmap.iter_ones(), self.signatures.iter())
+        {
+            signature.verify(&self.public_keys.public_keys[idx], message)?;
+        }
+        Ok(())
+    }
+
+    pub fn public_key_bytes(&self) -> Vec<u8> {
+        self.public_keys.to_bytes()
+    }
+
+    pub fn signature_bytes(&self) -> Vec<u8> {
+        bcs::to_bytes(&(&self.signatures, &self.signatures_bitmap))
+            .expect("Only unhandleable errors happen here.")
```

Comment:
This is thrown everywhere in the codebase... Maybe we should just have a expected to bytes for bcs.

Reply:
Would be a good idea...